neo/core/spiketrainlist.py

```diff
@@ -0,0 +1,150 @@
+        """Initialize self"""
+        self._items = items
+        self._spike_time_array = None
+        self._channel_id_array = None
+        self._all_channel_ids = None
+        self._spiketrain_metadata = None
+        self.segment = segment
+
+    def __iter__(self):
+        """Implement iter(self)"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        for item in self._items:
+            yield item
+
+    def __getitem__(self, i):
+        """x.__getitem__(y) <==> x[y]"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        return self._items[i]
+
+    def __str__(self):
+        """Return str(self)"""
+        if self._items is None:
+            if self._spike_time_array is None:
+                return str([])
+            else:
+                return "SpikeTrainList containing {} spikes from {} neurons".format(
+                    self._spike_time_array.size,
+                    self._channel_id_array.size)
+        else:
+            return str(self._items)
+
+    def __len__(self):
+        """Return len(self)"""
+        if self._items is None:
+            if self._all_channel_ids is not None:
+                return len(self._all_channel_ids)
+            elif self._channel_id_array is not None:
+                return np.unique(self._channel_id_array).size
+            else:
+                return 0
+        else:
+            return len(self._items)
+
+    def __add__(self, other):
+        """Return self + other"""
+        if isinstance(other, self.__class__):
+            if self._items is None or other._items is None:
+                # todo: update self._spike_time_array, etc.
+                raise NotImplementedError
+            else:
+                self._items.extend(other._items)
+            return self
+        elif other and isinstance(other[0], SpikeTrain):
+            for obj in other:
+                obj.segment = self.segment
+            self._items.extend(other)
+            return self
+        else:
+            return self._items + other
+
+    def __radd__(self, other):
+        """Return other + self"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        other.extend(self._items)
+        return other
+
+    def append(self, obj):
+        """L.append(object) -> None -- append object to end"""
+        if not isinstance(obj, SpikeTrain):
+            raise ValueError("Can only append SpikeTrain objects")
+        if self._items is None:
+            self._spiketrains_from_array()
+        obj.segment = self.segment
+        self._items.append(obj)
+
+    def extend(self, iterable):
+        """L.extend(iterable) -> None -- extend list by appending elements from the iterable"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        for obj in iterable:
+            obj.segment = self.segment
+        self._items.extend(iterable)
```

Comment:
maybe it's necessary to also verify that the items of `iterable` are instances of `SpikeTrain`? Otherwise arbitrary objects might be in the SpikeTrainList...

Reply:
yes, more robust checks are needed